src/goto-symex/field_sensitivity.cpp

```diff
@@ -0,0 +1,346 @@
+
+    mp_integer array_size;
+    if(to_integer(type.size(), array_size))
+      assert(false);
+    unsigned array_size_u=integer2unsigned(array_size);
+
+    dest=array_exprt(type);
+    dest.reserve_operands(array_size_u);
+    index_exprt index;
+    index.array()=array;
+    index.type()=type.subtype();
+
+    for(unsigned i=0; i<array_size_u; ++i)
+    {
+      index.index()=from_integer(i, index_type());
+
+      ssa_exprt tmp=ssa_expr;
+
+      tmp.type()=index.type();
+      tmp.set(ID_expression, index);
+      tmp.update_identifier();
+
+      exprt tmp_dest;
+      get_fields(ns, tmp, tmp_dest);
+      dest.copy_to_operands(tmp_dest);
+    }
+  }
+  else
+#endif
+    dest=ssa_expr;
+}
+
+/*******************************************************************\
+
+Function: field_sensitivityt::field_assignments
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+void field_sensitivityt::field_assignments(
+  const namespacet &ns,
+  goto_symex_statet &state,
+  symex_targett &target,
+  const exprt &lhs) const
+{
+  exprt lhs_fs=lhs;
+  operator()(ns, lhs_fs, false);
+
+  field_assignments_rec(ns, state, target, lhs_fs, lhs);
+}
+
+/*******************************************************************\
+
+Function: field_sensitivityt::field_assignments_rec
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+void field_sensitivityt::field_assignments_rec(
+  const namespacet &ns,
+  goto_symex_statet &state,
+  symex_targett &target,
+  const exprt &lhs_fs,
+  const exprt &lhs) const
+{
+  const typet &followed_type=ns.follow(lhs.type());
+
+  if(lhs==lhs_fs)
+    return;
+  else if(lhs_fs.id()==ID_symbol &&
+          lhs_fs.get_bool(ID_C_SSA_symbol))
+  {
+    exprt ssa_rhs=lhs;
+
+    state.rename(ssa_rhs, ns);
+    simplify(ssa_rhs, ns);
+
+    ssa_exprt ssa_lhs=to_ssa_expr(lhs_fs);
+    state.assignment(ssa_lhs, ssa_rhs, ns, true, true);
+
+    // do the assignment
+    target.assignment(
+      state.guard.as_expr(),
+      ssa_lhs,
+      ssa_lhs, ssa_lhs.get_original_expr(),
+      ssa_rhs,
+      state.source,
+      symex_targett::STATE);
+  }
+  else if(followed_type.id()==ID_struct)
+  {
+    const struct_typet &type=to_struct_type(followed_type);
+
+    const struct_union_typet::componentst &components=
+      type.components();
+
+    assert(components.empty() ||
+           (lhs_fs.has_operands() &&
+            lhs_fs.operands().size()==components.size()));
+
+    unsigned number=0;
+    for(struct_union_typet::componentst::const_iterator
+        it=components.begin();
+        it!=components.end();
+        ++it, ++number)
+    {
+      member_exprt member_rhs(lhs, it->get_name(), it->type());
+      exprt member_lhs=lhs_fs.operands()[number];
+
+      field_assignments_rec(ns, state, target, member_lhs, member_rhs);
+    }
+  }
+  else if(followed_type.id()==ID_array)
```

Comment:
Worth retaining the ability to represent a whole-array assignment compactly? String constants are the main thing that springs to mind -- then break it down into smaller pieces once the program starts making individual array cell assignments?